src/heat/heat_solve.cpp

```diff
@@ -73,20 +67,18 @@ double* heat_conduction_solver(
 }
 
 template <typename T>
-std::vector<double> linspace(T start_in, T end_in, int num_in)
-{
+std::vector<double> linspace(T start_in, T end_in, int num_in) {
     // Adapted from @Akavall
     // https://stackoverflow.com/questions/27028226/python-linspace-in-c
-    std::vector<double> linspaced;
-
-    double start = static_cast<double>(start_in);
-    double end = static_cast<double>(end_in);
-    double num = static_cast<double>(num_in);
+    std::vector<double> linspaced{};
 
-    if (num < 0)
-        throw std::invalid_argument(
-            "linspace: num_in must be a positive integer");
+    auto start = static_cast<double>(start_in);
```

Comment:
In general I think 'auto' makes it easier to write code, but the code is then harder to read. I think it's more important to make code easy to read. There are some good uses of auto, such as hiding ugly templating syntax which is basically language noise, such as std-library templated iterators etc. But usually I like to see the actual type rather than auto.
I guess in later c++ standards you can also use auto for function arguments, as kind of an alternative to templating. That's a separate issue.
In the code here, I think it would be more clear to simply write `double start = start_in;`. If the static_cast is safe, the assignment-conversion should be just as safe. (Unless it is doing a narrowing or sign-change that you want to explicitly do...)

Reply:
@rafmudaf First of all, I apologize for not explaining my reasoning for refactoring the code. Please find it as follows, which should answer your question as well.

[It is considered good practice](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#es11-use-auto-to-avoid-redundant-repetition-of-type-names) to use `auto` to avoid _redundant repetition_ of type names. For example

`double start = static_cast<double>(start_in);`

here the explicit type information on the left hand side is redundant (since it is clear we are casting it as a `double` from the rvalue), and we should prefer to use `auto` instead.

As a relevant aside, it's considered to be best practice to use [{}-initializer syntax for declaration with a named type](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#es23-prefer-the--initializer-syntax) to avoid unwanted (and implicit) narrowing conversions. However, with auto, we tend to use `=` because there no potentially dangerous type conversion involved.

@alanw0  Based on the discussion [here](https://stackoverflow.com/questions/1609163/what-is-the-difference-between-static-cast-and-c-style-casting), looks like the C++ style casts are better than C-style assignment conversion since the former is checked by the compiler and expresses the intent of the programmer. I am open to your suggestion on this.

Reply:
I agree that auto is used to avoid redundant repetition of types. But note in the good practices document you linked, it says 'long hard-to-remember types', and also specifies an exception to avoid using auto for '... cases where you know exactly which type you want' and of course in cases where an unintended initializer-list might result. In the end I guess it's a judgment call. I definitely think the motivating priority is readability rather than writability. Many times the use of auto means that someone else reading the code will have to 'dive' or look somewhere else to find out what the type actually is.

Regarding the conversion, I definitely agree that C++ casts are better than C-style casts. I'm not as sure about assignment conversions. I think that if the conversion will result in narrowing or other things, the compiler will indeed tell you. (At least it will if you have warnings on.) If you want to be as explicit as possible, then I agree that static_cast provides that. In this case since `start_in` is a template parameter, and thus is not assured of being a convertible type, then the static_cast is probably a good idea.
A common example of static_cast is to do something like `unsigned i = static_cast<unsigned>(-1);` which is legal and correct, but I think it's much more clear to do this: `unsigned i = std::numeric_limits<unsigned>::max();`. Of course in both cases you could use auto at the beginning, but I think the repetition is less of an issue for primitive types.

Reply:
> I definitely think the motivating priority is readability rather than writability.

Thanks for your comment, @alanw0 - I can definitely get behind your take above. Let's consider this comment resolved.

Reply:
Yes this is all good. Sorry for the delay in getting back to you. We are in good agreement. In practice if you use auto more than I would, I wouldn't hold up a pull-request over that.

Reply:
Thanks for the review, @alanw0!